### ISSUE TITLE
Toast is not shown during first trigger

### ISSUE THREAD

**Issue by igneqas** [iTwin/iTwinUI#1005]
#### Describe the bug (actual behavior)
After upgrading to iTwinUI v2, toasts do not show up the first time they are triggered but it works fine from the second time (only in our project). I tried recreating the problem in a new project but it works correctly there.

#### Expected behavior
Toast should be visible the first time it is triggered.

#### Reproduction
Could not be reproduced on a new project.

#### Additional information
iTwinUI - 2.2.0
React - 18

We did not modify our code for toasts after upgrading to v2. Also, the code is executed and toast is being created during the first trigger but it just doesn't show up. Our code is simple:
```
const toastSuccess = (message: string) => {
    toaster.closeAll();
    toaster.positive(message, { type: "temporary" });
  };
```
I guess this is some kind of timing issue, because, for example, if I add a 2 second sleep function between toaster.closeAll() and toaster.positive() it does show up during the first trigger (or calling toaster.closeAll() during the first render of the app).

**Comment by mayank99** [comment]
We talked offline and this seems to be because of react 18 batching updates. Wrapping `toaster.closeAll` in [`flushSync`](https://beta.reactjs.org/reference/react-dom/flushSync) fixes it.

I thought about maybe adding `flushSync` in our toaster code but decided against it because of how much react discourages and because this issue is not common enough.

If we get more users who face this issue in the future, we can revisit this.

**Comment by veekeys** [comment]
I think we also had this issue in our project, but solution added was to call `setSettings`... I would need to double check, if there is any more info I can gather on this. (react 18 with iTwinUI v1)

**Comment by mayank99** [comment]
> solution added was to call `setSettings`

that's really strange. it might have something to do with the async nature of `asyncInit` (which is being called in `setSettings` and `createToast`). i wonder if should make those functions `async` so that you can await them in application code like this:

```jsx
(async () => {
  await toaster.closeAll();
  await toaster.positive();
  await toaster.informational();
})();
```

**Comment by veekeys** [comment]
@mayank99 Give me a day or so as its busy week, I will try to test it in the code we have faced this issue!

**Comment by igneqas** [comment]
@mayank99 Hey, it would be great if you could create a snapshot for me, thanks.

**Comment by mayank99** [comment]
Released it in the `dev` channel. You can try it with `npm i @itwin/itwinui-react@dev` or pin the exact version:
```
"@itwin/itwinui-react": "2.3.1-dev.0"
```

**Comment by igneqas** [comment]
I tested the new changes and it didn't change anything. The problem is still there.

**Comment by mayank99** [comment]
That's unfortunate. well, i can't repro and least there's two workarounds now, so i'll close this issue.

I expect that at some point we will refactor the toaster anyway.

**Comment by mayank99** [comment]
fyi, all async behavior has been removed in #1351 which will be released in 3.0.

this snippet now works without hacks:
```jsx
const toaster = useToaster();

<Button onClick={() => toaster.positive('hello')}>Show</Button>
<Button
  onClick={() => {
    toaster.closeAll();
    toaster.positive('hello');
  }}
>
  Close all
</Button>
``` 

https://github.com/iTwin/iTwinUI/assets/9084735/c636b777-8bd5-4758-83ac-387186f2b0ed
